internal/security/secret.go

```diff
@@ -54,68 +54,73 @@ func NewSecretProvider(loggingClient logger.LoggingClient, configuration *common
+	ctx context.Context,
+	secretStoreInfo common.SecretStoreInfo) (pkg.SecretClient, error) {
+	var secretClient pkg.SecretClient
+
+	// secretStoreInfo is optional so that secret config can be empty
+	secretConfig, secretStoreEmpty, err := s.getSecretConfig(secretStoreInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// no secret client to be created
+	if secretStoreEmpty {
+		return nil, nil
+	}
+
 	if s.isSecurityEnabled() {
-		for i := 0; i < sharedSecretConfig.AdditionalRetryAttempts; i++ {
-			// create secret client based on SecretStore config for db credentials
-			s.SharedSecretClient, err = client.NewVault(ctx, sharedSecretConfig, s.loggingClient).Get(s.configuration.SecretStore)
+		for i := 0; i < secretConfig.AdditionalRetryAttempts; i++ {
+			secretClient, err = client.NewVault(ctx, secretConfig, s.loggingClient).Get(s.configuration.SecretStoreExclusive)
 			if err == nil {
 				break
 			} else {
-				waitTIme, err := time.ParseDuration(sharedSecretConfig.RetryWaitPeriod)
+				waitTIme, err := time.ParseDuration(secretConfig.RetryWaitPeriod)
 				if err != nil {
-					s.loggingClient.Error(fmt.Sprintf("invalid retry wait period for shared secret store config: %s", err.Error()))
-					return false
+					return nil, fmt.Errorf("invalid retry wait period for secret config: %s", err.Error())
 				}
 				time.Sleep(waitTIme)
 				continue
 			}
 		}
-		if err != nil {
-			s.loggingClient.Error(fmt.Sprintf("unable to create shared SecretClient: %s", err.Error()))
-			return false
-		}
 
-		for i := 0; i < exclusiveSecretConfig.AdditionalRetryAttempts; i++ {
-			// create secret client based on SecretStoreExclusive config for per-service credentials
-			s.ExclusiveSecretClient, err = client.NewVault(ctx, exclusiveSecretConfig, s.loggingClient).Get(s.configuration.SecretStoreExclusive)
-			if err == nil {
-				break
-			} else {
-				waitTIme, err := time.ParseDuration(exclusiveSecretConfig.RetryWaitPeriod)
-				if err != nil {
-					s.loggingClient.Error(fmt.Sprintf("invalid retry wait period for exlusive secret store config: %s", err.Error()))
-					return false
-				}
-				time.Sleep(waitTIme)
-				continue
-			}
-		}
 		if err != nil {
-			s.loggingClient.Error(fmt.Sprintf("unable to create exclusive SecretClient: %s", err.Error()))
-			return false
+			return nil, err
 		}
 	}
-	return true
+
+	return secretClient, nil
 }
 
 // getSecretConfig creates a SecretConfig based on the SecretStoreInfo configuration properties.
 // If a tokenfile is present it will override the Authentication.AuthToken value.
-func (s *SecretProvider) getSecretConfig(secretStoreInfo common.SecretStoreInfo) (vault.SecretConfig, error) {
+// the return boolean is used to indicate whether the secret store configuration is empty or not
+func (s *SecretProvider) getSecretConfig(secretStoreInfo common.SecretStoreInfo) (vault.SecretConfig, bool, error) {
+	emptySecretStore := common.SecretStoreInfo{}
+	if secretStoreInfo == emptySecretStore {
```

Comment:
Does this do a deep compare??

Reply:
yes, see the example here: https://play.golang.org/p/OvVR1ltisE5